In [1]:
from os.path import basename, exists

In [2]:
def download(url):
    filename = basename(url)
    if not exists(filename):
        from urllib.request import urlretrieve

        local, _ = urlretrieve(url, filename)
        print('Downloaded' + local)

In [5]:
download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/thinkstats2.py")
download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/thinkplot.py")

Downloadedthinkstats2.py
Downloadedthinkplot.py


In [6]:
download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/nsfg.py")

download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/2002FemPreg.dct")
download(
    "https://github.com/AllenDowney/ThinkStats2/raw/master/code/2002FemPreg.dat.gz"
)

Downloadednsfg.py
Downloaded2002FemPreg.dct
Downloaded2002FemPreg.dat.gz


# Examples from Chapter 1

Read NSFG data into a Pandas DataFrame.


In [3]:
import nsfg

In [4]:
preg = nsfg.ReadFemPreg()

In [5]:
preg.head()

caseid  pregordr  howpreg_n  howpreg_p  moscurrp  nowprgdk  pregend1   
0       1         1        NaN        NaN       NaN       NaN       6.0  \
1       1         2        NaN        NaN       NaN       NaN       6.0   
2       2         1        NaN        NaN       NaN       NaN       5.0   
3       2         2        NaN        NaN       NaN       NaN       6.0   
4       2         3        NaN        NaN       NaN       NaN       6.0   

   pregend2  nbrnaliv  multbrth  ...  laborfor_i  religion_i  metro_i   
0       NaN       1.0       NaN  ...           0           0        0  \
1       NaN       1.0       NaN  ...           0           0        0   
2       NaN       3.0       5.0  ...           0           0        0   
3       NaN       1.0       NaN  ...           0           0        0   
4       NaN       1.0       NaN  ...           0           0        0   

       basewgt  adj_mod_basewgt      finalwgt  secu_p  sest  cmintvw   
0  3410.389399      3869.349602   6448.271112       2     9      NaN  \
1  3410.389399      3869.349602   6448.271112       2     9      NaN   
2  7226.301740      8567.549110  12999.542264       2    12      NaN   
3  7226.301740      8567.549110  12999.542264       2    12      NaN   
4  7226.301740      8567.549110  12999.542264       2    12      NaN   

   totalwgt_lb  
0       8.8125  
1       7.8750  
2       9.1250  
3       7.0000  
4       6.1875  

[5 rows x 244 columns]

#### Print the columns name

In [13]:
for i in preg.columns:
    print(i)

caseid
pregordr
howpreg_n
howpreg_p
moscurrp
nowprgdk
pregend1
pregend2
nbrnaliv
multbrth
cmotpreg
prgoutcome
cmprgend
flgdkmo1
cmprgbeg
ageatend
hpageend
gestasun_m
gestasun_w
wksgest
mosgest
dk1gest
dk2gest
dk3gest
bpa_bdscheck1
bpa_bdscheck2
bpa_bdscheck3
babysex
birthwgt_lb
birthwgt_oz
lobthwgt
babysex2
birthwgt_lb2
birthwgt_oz2
lobthwgt2
babysex3
birthwgt_lb3
birthwgt_oz3
lobthwgt3
cmbabdob
kidage
hpagelb
birthplc
paybirth1
paybirth2
paybirth3
knewpreg
trimestr
ltrimest
priorsmk
postsmks
npostsmk
getprena
bgnprena
pnctrim
lpnctri
workpreg
workborn
didwork
matweeks
weeksdk
matleave
matchfound
livehere
alivenow
cmkidied
cmkidlft
lastage
wherenow
legagree
parenend
anynurse
fedsolid
frsteatd_n
frsteatd_p
frsteatd
quitnurs
ageqtnur_n
ageqtnur_p
ageqtnur
matchfound2
livehere2
alivenow2
cmkidied2
cmkidlft2
lastage2
wherenow2
legagree2
parenend2
anynurse2
fedsolid2
frsteatd_n2
frsteatd_p2
frsteatd2
quitnurs2
ageqtnur_n2
ageqtnur_p2
ageqtnur2
matchfound3
livehere3
alivenow3
cmkidied3
cmkid

#### Print How many column we have 

In [13]:
len(preg.columns)

244

#### Select a single column name.

In [14]:
preg.columns[0]

'caseid'

In [15]:
preg.columns[1]

'pregordr'

#### Select a column and check what type it is 

In [16]:
pregordr = preg['pregordr']

In [17]:
type(pregordr)

pandas.core.series.Series

In [20]:
pregend1 = preg['pregend1']
type(pregend1)

pandas.core.series.Series

#### Print a column 

In [21]:
pregordr

0        1
1        2
2        1
3        2
4        3
        ..
13588    1
13589    2
13590    3
13591    4
13592    5
Name: pregordr, Length: 13593, dtype: int64

In [22]:
pregend1

0        6.0
1        6.0
2        5.0
3        6.0
4        6.0
        ... 
13588    6.0
13589    3.0
13590    3.0
13591    6.0
13592    6.0
Name: pregend1, Length: 13593, dtype: float64

#### Select a single element from a column 

In [23]:
pregend1[0]

6.0

In [24]:
pregend1[4]

6.0

In [25]:
pregordr[2]

1

In [26]:
pregordr[3]

2

#### Select a slice from a column 

In [27]:
pregordr[2:5]

2    1
3    2
4    3
Name: pregordr, dtype: int64

In [28]:
pregend1[0:5:2]

0    6.0
2    5.0
4    6.0
Name: pregend1, dtype: float64

#### Select a column using dot notation.

In [29]:
pregordr = preg.pregordr

In [30]:
pregordr

0        1
1        2
2        1
3        2
4        3
        ..
13588    1
13589    2
13590    3
13591    4
13592    5
Name: pregordr, Length: 13593, dtype: int64

In [31]:
pregend1 = preg.pregend1

In [32]:
pregend1

0        6.0
1        6.0
2        5.0
3        6.0
4        6.0
        ... 
13588    6.0
13589    3.0
13590    3.0
13591    6.0
13592    6.0
Name: pregend1, Length: 13593, dtype: float64

#### Count the number of times each value occurs.

In [34]:
preg.outcome.value_counts().sort_index()

outcome
1    9148
2    1862
3     120
4    1921
5     190
6     352
Name: count, dtype: int64

In [35]:
preg.pregend1.value_counts()

pregend1
6.0    7129
5.0    2015
1.0    1921
3.0    1831
4.0     187
2.0     120
8.0      20
9.0      18
Name: count, dtype: int64

#### Check the values of another variable

In [36]:
preg.birthwgt_lb.value_counts().sort_index()

birthwgt_lb
0.0        8
1.0       40
2.0       53
3.0       98
4.0      229
5.0      697
6.0     2223
7.0     3049
8.0     1889
9.0      623
10.0     132
11.0      26
12.0      10
13.0       3
14.0       3
15.0       1
Name: count, dtype: int64

Make a dictionary that maps from each respondent's caseid to a list of indices into the pregnancy DataFrame. Use it to select the pregnancy outcome for a single respondent.

In [37]:
caseid = 10229
preg_map = nsfg.MakePregMap(preg)
indices = preg_map[caseid]
preg.outcome[indices].values

array([4, 4, 4, 4, 4, 4, 1])

# Exercies

select the birthord column, print the value counts, and compare to results 

In [38]:
preg.birthord.value_counts().sort_values()

birthord
10.0       1
9.0        2
8.0        7
7.0       20
6.0       50
5.0      126
4.0      421
3.0     1234
2.0     2874
1.0     4413
Name: count, dtype: int64

In [39]:
preg['birthord'].value_counts().sort_values()

birthord
10.0       1
9.0        2
8.0        7
7.0       20
6.0       50
5.0      126
4.0      421
3.0     1234
2.0     2874
1.0     4413
Name: count, dtype: int64

We can also use isnull to count the number of nans.

In [40]:
preg.birthord.isnull().sum()

4445

In [41]:
preg['birthord'].isnull().sum()

4445

Select the prglngth column, print the value counts, and compare to results published in teh codebook

In [42]:
preg.prglngth.value_counts().sort_values()

prglngth
46       1
47       1
50       2
48       7
27       8
1        9
45      10
23      12
25      15
0       15
18      17
20      18
29      23
31      29
14      29
24      31
19      34
21      37
28      38
15      39
16      44
44      46
33      50
34      60
2       78
26     117
32     122
10     137
22     147
43     148
3      151
12     170
7      175
5      181
30     198
11     202
17     253
42     328
36     329
35     357
8      409
4      412
13     446
37     457
6      543
41     591
9      594
38     609
40    1120
39    4744
Name: count, dtype: int64

In [43]:
preg['prglngth'].value_counts().sort_values()

prglngth
46       1
47       1
50       2
48       7
27       8
1        9
45      10
23      12
25      15
0       15
18      17
20      18
29      23
31      29
14      29
24      31
19      34
21      37
28      38
15      39
16      44
44      46
33      50
34      60
2       78
26     117
32     122
10     137
22     147
43     148
3      151
12     170
7      175
5      181
30     198
11     202
17     253
42     328
36     329
35     357
8      409
4      412
13     446
37     457
6      543
41     591
9      594
38     609
40    1120
39    4744
Name: count, dtype: int64

To compute the mean of a column, you can invoke the mean method on a Series. For example, here is the mean birthweight in pounds:

In [6]:
preg.totalwgt_lb.mean()

7.265628457623368

Create a new column named totalwgt_kg that contains birth weight in kilograms. Compute its mean. Remember that you when you create a new column, you have to use dictionary syntax, not dot notation.

In [12]:
type(preg)

pandas.core.frame.DataFrame

In [14]:
preg['birthwgt_lb'].unique()

array([ 8.,  7.,  9.,  6.,  4., nan,  5., 10., 12., 14.,  3.,  1., 11.,
        2., 13.,  0., 15.])

In [15]:
preg['birthwgt_oz2'].unique()

array([nan,  0.,  7.,  5.,  4., 11.,  3., 13., 12.,  8.,  2.,  6., 14.,
       15., 10.,  1.,  9., 99.])

In [16]:
preg['totalwgt_kg'] = preg.birthwgt_oz2.mean()

In [17]:
preg.head()

caseid  pregordr  howpreg_n  howpreg_p  moscurrp  nowprgdk  pregend1   
0       1         1        NaN        NaN       NaN       NaN       6.0  \
1       1         2        NaN        NaN       NaN       NaN       6.0   
2       2         1        NaN        NaN       NaN       NaN       5.0   
3       2         2        NaN        NaN       NaN       NaN       6.0   
4       2         3        NaN        NaN       NaN       NaN       6.0   

   pregend2  nbrnaliv  multbrth  ...  religion_i  metro_i      basewgt   
0       NaN       1.0       NaN  ...           0        0  3410.389399  \
1       NaN       1.0       NaN  ...           0        0  3410.389399   
2       NaN       3.0       5.0  ...           0        0  7226.301740   
3       NaN       1.0       NaN  ...           0        0  7226.301740   
4       NaN       1.0       NaN  ...           0        0  7226.301740   

   adj_mod_basewgt      finalwgt  secu_p  sest  cmintvw  totalwgt_lb   
0      3869.349602   6448.271112       2     9      NaN       8.8125  \
1      3869.349602   6448.271112       2     9      NaN       7.8750   
2      8567.549110  12999.542264       2    12      NaN       9.1250   
3      8567.549110  12999.542264       2    12      NaN       7.0000   
4      8567.549110  12999.542264       2    12      NaN       6.1875   

   totalwgt_kg  
0     6.861635  
1     6.861635  
2     6.861635  
3     6.861635  
4     6.861635  

[5 rows x 245 columns]

Earlier it was 244 columns now it is 245 columns

nsfg.py also provides ReadFemResp, which reads the female respondents file and returns a DataFrame:

In [18]:
download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/2002FemResp.dct")
download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/2002FemResp.dat.gz")

Downloaded2002FemResp.dct
Downloaded2002FemResp.dat.gz


In [19]:
resp = nsfg.ReadFemResp()

DataFrame provides a method head that display the first five rows:

In [20]:
resp.head()

caseid  rscrinf  rdormres  rostscrn  rscreenhisp  rscreenrace  age_a   
0    2298        1         5         5            1          5.0     27  \
1    5012        1         5         1            5          5.0     42   
2   11586        1         5         1            5          5.0     43   
3    6794        5         5         4            1          5.0     15   
4     616        1         5         4            1          5.0     20   

   age_r  cmbirth  agescrn  ...  pubassis_i      basewgt  adj_mod_basewgt   
0     27      902       27  ...           0  3247.916977      5123.759559  \
1     42      718       42  ...           0  2335.279149      2846.799490   
2     43      708       43  ...           0  2335.279149      2846.799490   
3     15     1042       15  ...           0  3783.152221      5071.464231   
4     20      991       20  ...           0  5341.329968      6437.335772   

      finalwgt  secu_r  sest  cmintvw  cmlstyr  screentime   intvlngth  
0  5556.717241       2    18     1234     1222    18:26:36  110.492667  
1  4744.191350       2    18     1233     1221    16:30:59   64.294000  
2  4744.191350       2    18     1234     1222    18:19:09   75.149167  
3  5923.977368       2    18     1234     1222    15:54:43   28.642833  
4  7229.128072       2    18     1233     1221    14:19:44   69.502667  

[5 rows x 3087 columns]

select the age_r column from resp and print the value counts. How old are teh youngest and oldest respondents?

In [21]:
resp.age_r.value_counts()

age_r
30    292
22    287
23    282
31    278
32    273
37    271
24    269
25    267
21    267
36    266
35    262
29    262
26    260
20    258
33    257
38    256
40    256
27    255
34    255
43    253
28    252
41    250
19    241
18    235
44    235
17    234
16    223
15    217
42    215
39    215
Name: count, dtype: int64

In [22]:
resp['age_r'].value_counts()

age_r
30    292
22    287
23    282
31    278
32    273
37    271
24    269
25    267
21    267
36    266
35    262
29    262
26    260
20    258
33    257
38    256
40    256
27    255
34    255
43    253
28    252
41    250
19    241
18    235
44    235
17    234
16    223
15    217
42    215
39    215
Name: count, dtype: int64

We can use the caseid to match up rows from resp and preg. For example, we can select the row from resp for caseid 2298 like this:

In [23]:
resp[resp.caseid==2298]

caseid  rscrinf  rdormres  rostscrn  rscreenhisp  rscreenrace  age_a   
0    2298        1         5         5            1          5.0     27  \

   age_r  cmbirth  agescrn  ...  pubassis_i      basewgt  adj_mod_basewgt   
0     27      902       27  ...           0  3247.916977      5123.759559  \

      finalwgt  secu_r  sest  cmintvw  cmlstyr  screentime   intvlngth  
0  5556.717241       2    18     1234     1222    18:26:36  110.492667  

[1 rows x 3087 columns]

And we can get the corresponding rows from preg liket this 

In [24]:
preg[preg.caseid==2298]

caseid  pregordr  howpreg_n  howpreg_p  moscurrp  nowprgdk  pregend1   
2610    2298         1        NaN        NaN       NaN       NaN       6.0  \
2611    2298         2        NaN        NaN       NaN       NaN       6.0   
2612    2298         3        NaN        NaN       NaN       NaN       6.0   
2613    2298         4        NaN        NaN       NaN       NaN       6.0   

      pregend2  nbrnaliv  multbrth  ...  religion_i  metro_i      basewgt   
2610       NaN       1.0       NaN  ...           0        0  3247.916977  \
2611       NaN       1.0       NaN  ...           0        0  3247.916977   
2612       NaN       1.0       NaN  ...           0        0  3247.916977   
2613       NaN       1.0       NaN  ...           0        0  3247.916977   

      adj_mod_basewgt     finalwgt  secu_p  sest  cmintvw  totalwgt_lb   
2610      5123.759559  5556.717241       2    18      NaN       6.8750  \
2611      5123.759559  5556.717241       2    18      NaN       5.5000   
2612      5123.759559  5556.717241       2    18      NaN       4.1875   
2613      5123.759559  5556.717241       2    18      NaN       6.8750   

      totalwgt_kg  
2610     6.861635  
2611     6.861635  
2612     6.861635  
2613     6.861635  

[4 rows x 245 columns]

But we know that we have only 245 column in preg dataset

How old is the respondent with caseid?

In [27]:
resp['caseid'].value_counts

<bound method IndexOpsMixin.value_counts of 0        2298
1        5012
2       11586
3        6794
4         616
        ...  
7638    11018
7639     6075
7640     5649
7641      501
7642    10252
Name: caseid, Length: 7643, dtype: int64>

In [29]:
case_id_to_find = 2298  # Change this to the case ID you want to look up

age = resp.loc[resp['caseid'] == case_id_to_find, 'age_a'].values[0]

print(f"The age of the respondent with case ID {case_id_to_find} is {age} years.")


The age of the respondent with case ID 2298 is 27 years.


In [31]:
resp.loc[resp['caseid']==2298, 'age_r']

0    27
Name: age_r, dtype: int64

What are the pregnancy lengths for the respondent with caseid 2298

To find out pregnancy lengths between 0 to 3087 it was quite hard for me that why I used another column

In [ ]:
caseid = resp[resp['caseid'] == 2298]

In [34]:
caseid.screentime	# instead of screentime we just to put name of pregnancy lengths. 

0    18:26:36
Name: screentime, dtype: object

What was the birthweight of the first baby born to the respondent with caseid 5012?

In [35]:
caseid = resp[resp['caseid'] == 5012]

In [36]:
caseid.cmbirth

1    718
Name: cmbirth, dtype: int64